Added example that shows text generation with miniature GPT

apoorvnandan opened the pull request:
This example demonstrates autoregressive language modeling and text generation using this more manageable version of the GPT, written from scratch in `tf.keras`.

I know the amazing results GPT-2 and GPT-3 have shown are largely due to the scale at which they operate. But, this miniature version also starts generating coherent movie reviews after 30 or so epochs. (Each epoch is ~30 seconds on TPU)

fchollet commented on examples/nlp/text_generation_with_miniature_gpt.py:
```diff
@@ -0,0 +1,331 @@
+        self.dropout2 = layers.Dropout(rate)
+
+    def call(self, inputs, training):
+        attn_output = self.att(inputs)
```
Use fully-spelled out variable names

fchollet commented on examples/nlp/text_generation_with_miniature_gpt.py:
```diff
@@ -0,0 +1,331 @@
+    outputs = layers.Dense(vocab_size, activation="softmax")(x)
+    model = keras.Model(inputs=inputs, outputs=[outputs, x])
+    model.compile(
+        "adam", loss=["sparse_categorical_crossentropy", None],
```
Add a comment about the None loss for the 2-output model, since it won't be obvious to most readers

fchollet commented on examples/nlp/text_generation_with_miniature_gpt.py:
```diff
@@ -0,0 +1,331 @@
+        self, max_tokens, start_tokens, index_to_word, temp=1.0, print_every=5
+    ):
+        """
+        :param max_tokens: number of tokens to be generated
```
Use the same format for docstring as what's used in `tf.keras` docstrings

fchollet commented on examples/nlp/text_generation_with_miniature_gpt.py:
```diff
@@ -0,0 +1,331 @@
+
+"""
+## Create model
+"""
```
Add a comment about using TPU

fchollet commented on examples/nlp/text_generation_with_miniature_gpt.py:
```diff
@@ -0,0 +1,331 @@
+"""
+
+
+(x_train, y_train), (x_val, y_val) = keras.datasets.imdb.load_data(num_words=vocab_size)
```
Consider starting from the raw data and using the `TextVectorization` layer for tokenization (it already handles OOV and padding). The shifting can be done via `dataset.map()`. This would make the example much more generalizable since people could drop in their own folder of texts.

Ref: https://keras.io/examples/nlp/text_classification_from_scratch/

apoorvnandan replied:
Just noticed, the referred example doesn't work in google colab unless you do a `!pip install tf-nightly` and restart the runtime. Maybe there should be a warning line in that example till `text_dataset_from_directory` becomes a part of the stable release.

fchollet replied:
Yes, in general everything on keras.io expects tf-nightly at this time. These features will be in TF 2.3, which will be out within a month or so. So this issue is temporary.

apoorvnandan replied:
And Colab TPU throws an error with `tf-nightly`.

fchollet replied:
Please file a report, and try to install an earlier version of tf-nightly: https://pypi.org/project/tf-nightly/#history

fchollet commented on examples/nlp/text_generation_with_miniature_gpt.py:
```diff
@@ -0,0 +1,331 @@
+maxlen = 200  # Max sequence size
+embed_dim = 32  # Embedding size for each token
+num_heads = 2  # Number of attention heads
+ff_dim = 32  # Hidden layer size in feed forward network inside transformer
```
Use fully-spelled out variable names

apoorvnandan commented:
Thanks for the comments. Good to know about the upcoming `MultiHeadAttention` layer. I was wondering when that would become part of the core API.

I'll make the required changes.

apoorvnandan commented:
Updated the example.
- Loads data using `tf.data.TextLineDataset` from any list of text files and tokenizes with `TextVectorization` layer.
- Added comments and introduction section.

Some issues:
- When running on TPU, the session crashes with the message "Your session crasshed for an unknown reason" and I can't find any helpful logs. The same code works fine on GPU and CPU.
  GPU takes 145 seconds / epoch compared to 25 seconds on TPU.
- Black fails to format the bash commands for downloading the data.
```
Cannot parse: 207:0: !curl -O https://ai.stanford.edu/~amaas/data/sentiment/aclImdb_v1.tar.gz
```

fchollet commented:
> Black fails to format the bash commands for downloading the data.

Please put shells commands in shell blocks, like this:

```python
"""shell
!cmd
"""
```

You can examples of this in other tutobooks.

fchollet commented on examples/nlp/text_generation_with_miniature_gpt.py:
```diff
@@ -0,0 +1,355 @@
+# Tokenize starting prompt
+word_to_index = {}
+for index, word in enumerate(vocab):
+    word_to_index[word.decode('utf-8')] = index + 2  # PAD and OOV at index 0 and 1
```
We recently changed the `TextVectorization.get_vocabulary` API (it's an experimental API, so subject to improvements). It returns strings (so no need to decode) and it already includes padding and oov (so no need for the +2 offset).

apoorvnandan replied:
Yeah in one particular session, I was getting strings + pad and OOV tokens in the vocabulary. And then I restarted the session and got bytes 😅

fchollet replied:
Please use a recent tf-nightly version (so the example will be future-proof). It would be good to also add a comment to indicate that the example should be run with tf-nightly higher than [FILL ME] or with TF 2.3 or higher.

apoorvnandan replied:
Yes. Reinstalled tf-nightly (`2.3.0-dev20200531`) and everything is working smoothly except TPUs. The GPU is running pretty fast, so this example is very much doable without TPU now.

apoorvnandan replied:
The TPU error has been reported before ([31st march](https://github.com/tensorflow/tensorflow/issues/38069)). And apparently it was fixed in some versions but has reappeared now.

apoorvnandan replied:
With the stable release and some older versions of tf-nightly, this issue does not come up, but then the session crashes unexpectedly during model.fit

fchollet commented on examples/nlp/text_generation_with_miniature_gpt.py:
```diff
@@ -0,0 +1,355 @@
+
+class TextGenerator(keras.callbacks.Callback):
+    """
+    Callback to generate text from trained model.
```
Put this line on the first docstring line, e.g.

```
"""Callback to generate text from the trained model.
```

fchollet commented on examples/nlp/text_generation_with_miniature_gpt.py:
```diff
@@ -0,0 +1,355 @@
+        max_tokens: Integer, the number of tokens to be generated after prompt.
+        start_tokens: List of integers, the token indices for the starting prompt.
+        index_to_word: List of strings, obatined from TextVectorization layer.
+        temp: Float, dictates the variation during sampling.
```
Prefer using "temperature", "temp" is ambiguous since it is usually used for "temporary"

apoorvnandan commented:
The epoch time on GPU is down to ~44 seconds. :) (due to prefetch I think)

fchollet commented on examples/nlp/text_generation_with_miniature_gpt.py:
```diff
@@ -0,0 +1,358 @@
+
+- [GPT](https://s3-us-west-2.amazonaws.com/openai-assets/research-covers/language-unsupervised/language_understanding_paper.pdf)
+- [GPT-2](https://d4mucfpksywv.cloudfront.net/better-language-models/language_models_are_unsupervised_multitask_learners.pdf)
+- [GPT-3](https://arxiv.org/pdf/2005.14165.pdf)
```
Prefer linking to the abstract on ArXiv, not the PDF

fchollet commented on examples/nlp/text_generation_with_miniature_gpt.py:
```diff
@@ -0,0 +1,358 @@
+**References:**
+
+- [GPT](https://s3-us-west-2.amazonaws.com/openai-assets/research-covers/language-unsupervised/language_understanding_paper.pdf)
+- [GPT-2](https://d4mucfpksywv.cloudfront.net/better-language-models/language_models_are_unsupervised_multitask_learners.pdf)
```
Are these 2 papers on ArXiv? If so, link to the abstract on ArXiv.

apoorvnandan replied:
These are not on arxiv. They are there on semantic scholar though

apoorvnandan commented:
Should I remove TPU related comments and code or just leave it as it is?

fchollet commented:
If it isn't currently working, let's remove it. The example runs fast enough on GPU anyway.

apoorvnandan commented:
Made some bug fixes in the callback. (starting tokens were getting updated in every call)
Changed from temperature sampling to top K sampling. It reduces processing time and starts producing coherent text earlier. (It was used in GPT-2 paper as well)

Generated `md` and `ipynb` files.
Please check.

fchollet commented on scripts/tmp_6494311/text_generation_with_miniature_gpt.ipynb:
```diff
@@ -0,0 +1,539 @@
+{
```
Please remove the temporary files that were added.

fchollet commented on examples/nlp/text_generation_with_miniature_gpt.py:
```diff
@@ -0,0 +1,338 @@
+"""
+
+"""shell
+!curl -O https://ai.stanford.edu/~amaas/data/sentiment/aclImdb_v1.tar.gz
```
These commands should not start with `!`. The `!` is added automatically.

fchollet commented on examples/nlp/text_generation_with_miniature_gpt.py:
```diff
@@ -0,0 +1,338 @@
+
+
+text_ds = text_ds.map(prepare_lm_inputs_labels)
+text_ds = text_ds.repeat().prefetch(tf.data.experimental.AUTOTUNE)
```
Do not use `repeat`. Just pass a finite dataset to `fit`. Do not pass `steps_per_epoch`.

apoorvnandan commented:
Made the changes as per your review. 

I started tweaking parameters, trying different datasets etc to see why the generated text was not coherent. Then I noticed that incoherent text was generated only when running the code via `autogen.py`. And that was because that particular Colab session was using tf-nightly installed from `requirements.txt` which has the version `tf-nightly==2.2.0.dev20200508`. This version does not return PAD and UNK tokens in the vocabulary. So all decoded tokens in the generated text were pretty much random. 🤦 🤦 

Added comment that this example should be run with `tf-nightly>=2.3.0-dev20200531`. (or tf 2.3+)

I'm not sure how soon will these features become a part of the stable version. But I think a lot of people will be reading these examples before that, and they are likely to get confused and frustrated when they can't replicate them without issues on their end.

fchollet commented:
> I'm not sure how soon will these features become a part of the stable version. But I think a lot of people will be reading these examples before that, and they are likely to get confused and frustrated when they can't replicate them without issues on their end.

They will be in TF 2.3. The release branch will be cut in 2 weeks and the final release will be in 1-2 months.

We can pause the submission of this example until 2.3 is out, if you'd like.

apoorvnandan commented:
> The release branch will be cut in 2 weeks

That is sooner than expected. :)

The examples are totally fine. They show awesome functionalities like `TextVectorization` and `preprocessing.text_dataset_from_directory`. These simplify important and frequent workflows. 

But maybe there should be a little more attention to adding comments at the beginning of an example regarding tf versions. (Even if it's just for a few weeks)

In particular, code with `TextVectorization` can fail silently, which can be a bad experience. I would be happy to raise PRs for adding these comments to 
https://keras.io/examples/nlp/text_classification_from_scratch/
https://keras.io/examples/nlp/pretrained_word_embeddings/
And remove/update them once the stable release comes out. (in a few weeks)

Let me know if this feels like overthinking.

fchollet commented:
> In particular, code with TextVectorization can fail silently, which can be a bad experience. I would be happy to raise PRs for adding these comments to

Yes, please do, that would be very helpful.

fchollet commented on examples/nlp/md/text_generation_with_miniature_gpt.md:
```diff
@@ -0,0 +1,668 @@
+1563/1563 - 16s - loss: 4.2981 - dense_6_loss: 4.2981
+Epoch 8/30
+generated text:
+this movie is one of the worst movies i have ever seenbr br it
```
In general with raw IMDB data you should strip the `<br>` tags (like we do in the text classification example). Otherwise only the `<` and `>` characters get stripped and you end up with random "br" strings in your text

fchollet commented on examples/nlp/md/text_generation_with_miniature_gpt.md:
```diff
@@ -0,0 +1,668 @@
+print(f"{len(filenames)} files")
+
+# Create dataset from text files
+text_ds = tf.data.TextLineDataset(filenames)
```
You need to randomize the files. Your generated reviews are all 100% negative. That is because the negative reviews form the second 25% of the training data, and you generate at the end of each epoch. So by the time you generate, the last 12k or so reviews seen by the model were all negative. Hence the results.

apoorvnandan replied:
Sure, I'll add randomization.

fchollet commented on examples/nlp/md/text_generation_with_miniature_gpt.md:
```diff
@@ -0,0 +1,668 @@
+print(f"{len(filenames)} files")
+
+# Create dataset from text files
+text_ds = tf.data.TextLineDataset(filenames)
```
One thing that isn't clear to me with the use of `TextLineDataset` is this:
is one sample a *line* from a file, or the entire content of a file regardless of line breaks? I presume files may contain line breaks. It would be useful to print a few reviews from the dataset to check what they look like. Doing so would also have pointed you towards stripping the `<br>` tags in the first place.

apoorvnandan replied:
`TextLineDataset` creates concatenation of all the lines from all the files. If file 1 has 2 lines, and file 2 has 3 lines, the dataset from `[file1, file2]` will have 5 items. 
Incase of IMDB, all the files seem to have just one line. The line breaks are present in the form of `<br>` tag.
I noticed them but I thought its fine for demonstration purposes even if we keep them. Or if we print them as `\n`. For example, we are also ignoring periods and commas which makes input text somewhat meaningless by joining all the sentences.

fchollet commented on examples/nlp/md/text_generation_with_miniature_gpt.md:
```diff
@@ -0,0 +1,668 @@
+
+# Create vectcorization layer and adapt it to the text
+vectorize_layer = TextVectorization(
+    max_tokens=vocab_size - 1, output_mode="int", output_sequence_length=maxlen + 1,
```
why `maxlen + 1` and not `maxlen`?

apoorvnandan replied:
If the original line is `[a,b,c,d,e]` and `maxlen = 3` I would be processing it afterwards into two sequences. `input=[a,b,c], target=[b,c,d]`. In this step I just truncate it to `maxlen+1 = 4` so that I can create the input target pair effectively later on.
This is not ideal, but it's just for demonstration. If the review is long enough it would be best to keep all of it during tokenization and slide a window of size `maxlen` over it to create multiple data points.

fchollet commented on examples/nlp/md/text_generation_with_miniature_gpt.md:
```diff
@@ -0,0 +1,668 @@
+
+start_prompt = "this movie is"
+start_tokens = [word_to_index.get(_, 1) for _ in start_prompt.split()]
+num_tokens_generated = 10
```
Consider generating more; 10 words is very little to judge quality. 40-50 is better.

fchollet commented on examples/nlp/md/text_generation_with_miniature_gpt.md:
````diff
@@ -0,0 +1,668 @@
+
+```python
+vocab_size = 20000  # Only consider the top 20k words
+maxlen = 20  # Max sequence size
````
20 words is very little context. I would recommend 100 or so.

apoorvnandan replied:
This was done for speeding things up. I'll increase this and the generated text length, and check the results.

apoorvnandan replied:
When generating longer sentences, the text may lose some readability as the punctuations are being removed. So sentences won't be separated.

apoorvnandan commented:
- Added randomization. 
- Changed preprocessing to remove `<br>` tags and separate out punctuations instead of removing them.
- Increased context and generation lengths

fchollet commented on examples/nlp/md/text_generation_with_miniature_gpt.md:
```diff
@@ -0,0 +1,682 @@
+
+# Create dataset from text files
+text_ds = tf.data.TextLineDataset(filenames)
+text_ds = text_ds.shuffle(buffer_size=50000)
```
What I would recommend for shuffling is:

- Shuffle the filenames before creating the dataset, e.g. `random.shuffle(filenames)`
- Use a small `buffer_size` for dataset shuffling to do local shuffling at each epoch (typically 8x the batch size)

The larger the shuffle buffer the less performant your dataset will be.

This does not affect the results compared to the current code so this edit can be done without re-generating the files (it just need to be applied in all 3 places)

apoorvnandan replied:
Got it. I'll make the changes.

notnot commented on examples/nlp/md/text_generation_with_miniature_gpt.md:
```diff
@@ -0,0 +1,449 @@
+781/781 - 19s - loss: 4.1893 - dense_6_loss: 4.1893
+Epoch 15/15
+generated text:
+the film is br it from idea of that time one a show movie a show the the the the the the the the
```
I have noticed this as well, especially when outputting more that the default 40 words. The generator usually ends in in a repetitive loop.

notnot commented:
I'm seeing poor performance when using this code with other text data. It seems the model can't deal with newlines properly, because it get stuck generating '0' tokens after that. In the IMDB case, the lines are very long, this is probably why it works reasonably well. But even there, if you increase the number of worder to generate to say 100, you see that the transformer has a lot of problems, often getting stuck in repetitive loops.
